examples/src/bin/gif_paintable.rs

```diff
@@ -0,0 +1,305 @@
+        type Type = super::GifPaintable;
+        type ParentType = glib::Object;
+        type Instance = subclass::simple::InstanceStruct<Self>;
+        type Class = subclass::simple::ClassStruct<Self>;
+
+        glib::object_subclass!();
+
+        fn type_init(type_: &mut subclass::InitializingType<Self>) {
+            type_.add_interface::<gdk::Paintable>();
+        }
+
+        fn new() -> Self {
+            Self {
+                next_frame: RefCell::new(None),
+                frames: RefCell::new(vec![]),
+                width: Cell::new(0),
+                height: Cell::new(0),
+                current_idx: Cell::new(0),
+            }
+        }
+    }
+
+    impl ObjectImpl for GifPaintable {
+        fn constructed(&self, obj: &Self::Type) {
+            self.parent_constructed(obj);
+        }
+
+        fn dispose(&self, _obj: &Self::Type) {}
+    }
+
+    impl PaintableImpl for GifPaintable {
+        fn get_intrinsic_height(&self, _paintable: &Self::Type) -> i32 {
+            self.height.get()
+        }
+
+        fn get_intrinsic_width(&self, _paintable: &Self::Type) -> i32 {
+            self.width.get()
+        }
+
+        fn snapshot(
+            &self,
+            _paintable: &Self::Type,
+            snapshot: &gdk::Snapshot,
+            width: f64,
+            height: f64,
+        ) {
+            if let Some(texture) = self.next_frame.borrow_mut().take() {
+                let w = self.width.get() as f64;
+                let h = self.height.get() as f64;
+                texture.snapshot(snapshot, w, h);
+            } else {
+                let snapshot = snapshot.downcast_ref::<gtk::Snapshot>().unwrap();
+                snapshot.append_color(
+                    &gdk::RGBA::black(),
+                    &graphene::Rect::new(0f32, 0f32, width as f32, height as f32),
+                );
+            }
+        }
+    }
+}
+
+glib::wrapper! {
+    pub struct GifPaintable(ObjectSubclass<imp::GifPaintable>) @implements gdk::Paintable;
+}
+
+impl GifPaintable {
+    pub fn new() -> Self {
+        glib::Object::new(&[]).expect("Failed to create object of type GifPaintable")
+    }
+
+    // Loads the bytes of a GIF into the paintable
+    // The loading consists of decoding the gif with a GIFDecoder
+    // Then storing the frames so that the paintable can render them
+    pub fn load_form_bytes(&self, bytes: Vec<u8>) -> Result<(), Box<dyn std::error::Error>> {
+        let self_ = imp::GifPaintable::from_instance(self);
+        self_.current_idx.set(0);
+
+        let read = Cursor::new(bytes);
+        let decoder = GifDecoder::new(read)?;
+        let frames = decoder
+            .into_frames()
+            .collect_frames()?
+            .into_iter()
+            .map(Frame::new)
+            .collect::<Vec<Frame>>();
+
+        self_.frames.replace(frames);
+        self.invalidate_contents();
+        glib::timeout_add_local(
+            std::time::Duration::from_millis(60),
```

Comment:
GIFs usually continue information about the framerate or rather when to schedule frames. Every 60ms is not going to be correct in general.

Reply:
Yes, that part is going to be cleaned up a bit, especially that we should stop the timeout at dispose